lib/coinbase/smart_contract.rb

```diff
@@ -126,6 +132,95 @@ def self.create_multi_token_contract(
+    #
+    # @param network [Coinbase::Network, Symbol] The Network or Network ID of the Asset
+    # @param contract_address [String] The address of the deployed contract
+    # @param method [String] The name of the method to call on the contract
+    # @param abi [Array, nil] The ABI of the contract. If nil, the method will attempt to use a cached ABI
+    # @param args [Hash] The arguments to pass to the contract method.
+    #   The keys should be the argument names, and the values should be the argument values.
+    # @return [Object] The result of the contract call, converted to an appropriate Ruby type
+    # @raise [Coinbase::ApiError] If there's an error in the API call
+    def self.read(
+      network:,
+      contract_address:,
+      method:,
+      abi: nil,
+      args: {}
+    )
+      network = Coinbase::Network.from_id(network)
+
+      response = Coinbase.call_api do
+        smart_contracts_api.read_contract(
+          network.normalized_id,
+          contract_address,
+          {
+            method: method,
+            args: (args || {}).to_json,
+            abi: abi&.to_json
+          }
+        )
+      end
+
+      convert_solidity_value(response)
+    end
+
+    # Converts a Solidity value to an appropriate Ruby type.
+    #
+    # @param solidity_value [Coinbase::Client::SolidityValue] The Solidity value to convert
+    # @return [Object] The converted Ruby value
+    # @raise [ArgumentError] If an unsupported Solidity type is encountered
+    #
+    # This method handles the following Solidity types:
+    # - Integers (uint8, uint16, uint32, uint64, uint128, uint256, int8, int16, int32, int64, int128, int256)
+    # - Address
+    # - String
+    # - Bytes (including fixed-size byte arrays)
+    # - Boolean
+    # - Array
+    # - Tuple (converted to a Hash)
+    #
+    # For complex types like arrays and tuples, the method recursively converts nested values.
+    def self.convert_solidity_value(solidity_value)
+      return nil if solidity_value.nil?
+
+      type = solidity_value.type
+      value = solidity_value.value
+      values = solidity_value.values
+
+      case type
+      when 'uint8', 'uint16', 'uint32', 'uint64', 'uint128', 'uint256',
+       'int8', 'int16', 'int32', 'int64', 'int128', 'int256'
+        value&.to_i
+      when 'address', 'string', /^bytes/
+        value
+      when 'bool'
+        if value.is_a?(String)
+          value == 'true'
+        else
+          !value.nil?
+        end
+      when 'array'
+        values ? values.map { |v| convert_solidity_value(v) } : []
+      when 'tuple'
+        if values
+          result = {}
+          values.each do |v|
+            raise ArgumentError, 'Error: Tuple value without a name' unless v.respond_to?(:name)
+
+            result[v.name] = convert_solidity_value(v)
+          end
+          result
+        else
+          {}
+        end
+      else
+        raise ArgumentError, "Unsupported Solidity type: #{type}"
```

Comment:
Would be good to add tests for the two invalid / ArgumentErrors as well, but not urgent/blocking

Reply:
good call! will track this